File: wrappers/julia/src/TurboToken.jl
module TurboToken

include("error.jl")
include("ffi.jl")
include("registry.jl")
include("rank_cache.jl")
include("chat.jl")
include("encoding.jl")

export get_encoding, get_encoding_for_model, list_encoding_names, version
export Encoding, encode, decode, count, count_tokens, is_within_token_limit
export encode_chat, count_chat, is_chat_within_token_limit
export encode_file_path, count_file_path, is_file_path_within_token_limit

"""
    version()

Return the turbotoken native library version string.
"""
version() = ffi_version()

"""
    get_encoding(name::AbstractString) -> Encoding

Get a BPE encoding by name (e.g. "cl100k_base", "o200k_base").
"""
function get_encoding(name::AbstractString)
    spec = get_encoding_spec(String(name))
    rank_payload = read_rank_file(spec.name)
    return Encoding(spec.name, spec, rank_payload)
end

"""
    get_encoding_for_model(model::AbstractString) -> Encoding

Get the appropriate BPE encoding for a given model name.
"""
function get_encoding_for_model(model::AbstractString)
    enc_name = model_to_encoding(String(model))
    return get_encoding(enc_name)
end

end # module

File: wrappers/ruby/lib/turbotoken.rb
require_relative "turbotoken/version"
require_relative "turbotoken/ffi_bridge"
require_relative "turbotoken/registry"
require_relative "turbotoken/rank_cache"
require_relative "turbotoken/encoding"
require_relative "turbotoken/chat"

module TurboToken
  class Error < StandardError; end

  # Get an encoding by name (e.g. "cl100k_base", "o200k_base").
  def self.get_encoding(name)
    spec = Registry.get_encoding_spec(name)
    rank_payload = RankCache.ensure_rank_file(name)
    Encoding.new(name, spec, rank_payload)
  end

  # Get the encoding for a given model name (e.g. "gpt-4o", "gpt-3.5-turbo").
  def self.get_encoding_for_model(model)
    encoding_name = Registry.model_to_encoding(model)
    get_encoding(encoding_name)
  end

  # List all supported encoding names.
  def self.list_encoding_names
    Registry.list_encoding_names
  end

  # Return the turbotoken native library version string.
  def self.version
    FFIBridge.turbotoken_version
  end
end
